eventlet/green/ssl.py

```diff
@@ -58,34 +57,48 @@ def __init__(self, sock, keyfile=None, certfile=None,
             self._timeout = sock.gettimeout()
 
         if sys.version_info >= (3, 7):
-            # Monkey-patch the sslsocket so our modified self gets
-            # injected into its _create method.
-            def fake_new(self, cls, *args, **kwargs):
-                return self
-
-            orig_new = _original_sslsocket.__new__
-            try:
-                _original_sslsocket.__new__ = fake_new.__get__(self, GreenSSLSocket)
-
-                self = _original_wrap_socket(
-                    sock=sock.fd,
-                    keyfile=keyfile,
-                    certfile=certfile,
-                    server_side=server_side,
-                    cert_reqs=cert_reqs,
-                    ssl_version=ssl_version,
-                    ca_certs=ca_certs,
-                    do_handshake_on_connect=False,
-                    *args, **kw
-                )
-                self.keyfile = keyfile
-                self.certfile = certfile
-                self.cert_reqs = cert_reqs
-                self.ssl_version = ssl_version
-                self.ca_certs = ca_certs
-            finally:
-                # Unpatch
-                _original_sslsocket.__new__ = orig_new
+            import imp
+            need_swap_ssl = 'ssl' in sys.modules
+            if need_swap_ssl:
+                # In case 'ssl' is greened, temporarily swap back original SSLSocket
+                # and SSLContext because Python 3.7 use this exact Name to locate the
+                # correct parent class in SSLSocket._create
+                original_sslcontext = _original_sslcontext
+                imp.acquire_lock()
```

Comment:
I confess that I know no more about the `imp` lock than what I quoted above. I agree that, absent a definite answer, it's safer to retain the locking than to remove it.

That said, is it necessary to hold the lock during the body of the `with` statement? Does `ssl.wrap_socket()` aka `_original_wrap_socket()` patch anything in `sys.modules`? Would it be just as safe to acquire the lock while patching, release it, execute the body of the `with` statement, reacquire the lock to restore the patched values and release it again?

I guess the problem case would be if some other thread imports something else that depends on `ssl` during that time: it would see the original `ssl` module rather than the patched one. Is that a fair statement? If so, it appears that holding the lock for the duration is still valuable. Yes?

Reply:
Added more details to the background section in the PR description.